### PR TITLE
[WasmFS] Reading and writing in the Node backend

### PR DESCRIPTION
**Pull request by tlively**
Implement file reading and writing in the Node WasmFS backend. Since there are
no backend hooks for determining when a file is opened or closed, we cannot
store native file descriptors for open files, so we instead open and close files
each time they are read or written.

Also fix the return value of _wasmfs_node_get_mode as a drive-by.

### CODE REVIEW

File: src/library_wasmfs_node.js [hunk 1]
```diff
@@ -67,10 +67,10 @@ mergeInto(LibraryManager.library, {
   _wasmfs_node_get_mode: function(path_p, mode_p) {
     let stat = wasmfsNodeLstat(UTF8ToString(path_p));
     if (stat === undefined) {
-      return 0;
+      return 1;
     }
-    {{{ makeSetValue('mode_p', 0, 'stat.mode', 'i32') }}}
-    return 1;
+    {{{ makeSetValue('mode_p', 0, 'stat.mode', 'i32') }}};
+    // return 0;
```
Comment (kripken) [inline src/library_wasmfs_node.js:73]:
```suggestion
    return 0;
```

File: system/lib/wasmfs/backends/node_backend.cpp [hunk 2]
```diff
@@ -41,22 +55,40 @@ class NodeFile : public DataFile {
...
-    WASMFS_UNREACHABLE("TODO: implement NodeFile::getSize");
-    return 0;
+    // TODO: This should really be using a 64-bit file size type rather than
+    // size_t.
+    uint32_t size;
+    if (_wasmfs_node_get_file_size(path.c_str(), &size)) {
+      // TODO: Make this fallible
+      return 0;
+    }
+    emscripten_console_logf("Got a size of %u for %s", size, path.c_str());
```
Comment (kripken) [inline system/lib/wasmfs/backends/node_backend.cpp:65]:
```suggestion
```

**Review by kripken — APPROVED** [review]
lgtm % comments

Should we think more about backend hooks for file open/close? Or can we make progress without that for now?

File: src/library_wasmfs_node.js (hunk 1 above)
Comment (tlively) [reply src/library_wasmfs_node.js:73]: Should we replace all the `// return 0;` in this file with actual returns? I've been depending on the implicit return of undefined converting to 0 in Wasm.

**Comment by tlively** [comment]
> Should we think more about backend hooks for file open/close? Or can we make progress without that for now?

We should think about this more, I think. It is important for the correctness of the unlink-then-access idiom. I'll work on this next.

File: src/library_wasmfs_node.js (hunk 1 above)
Comment (kripken) [reply src/library_wasmfs_node.js:73]: It is ok to do so, but in that case maybe the comment should be more explicit. I see we have this comment already in library.js:
```js
    // implicit 0 return to save code size (caller will cast "undefined" into 0
    // anyhow)
```
Perhaps `// implicit return 0` or `// return 0 implicitly`?

**Pull request merged**